### PR TITLE
SEXP library integration

### PR DESCRIPTION
**Pull request by maxirmx**

### CODE REVIEW

**Comment by maxirmx** [comment]
https://github.com/rnpgp/sexp/issues/1

**Comment by ni4** [comment]
@maxirmx Approach with `FetchContent_Declare` could/would be problematic for us as package maintainers may use systems without internet connection or even without the git, as we experienced with RNP alone. My suggestion would be to somehow incorporate libary source into the RNP tree (just copy? Git submodule?) and check whether it's up-to-date via CI.
@ronaldtse @ribose-jeffreylau Maybe you have some input on this?

**Comment by ronaldtse** [comment]
@maxirmx : @ni4 is correct that we need to take care of the following audience:
- package maintainers that may not have git or do not have internet access

In general, there needs to be a way to package an archive (tar.gz) that can be fully compiled without additional remote steps.

As part of the PR, can you also update documentation so that package maintainers can see how to compile RNP for their system with the sexp library? Thanks.

**Comment by maxirmx** [comment]
@ronaldtse @ni4 
When I have it integrated I will remove FetchContent and replace it with whatever is applicable. 
Thank you

**Comment by antonsviridenko** [comment]
I agree that build system should not download anything,
the most canonical way is to create separate package named like 'librnpsexp' or something like that and make RNP packages (optionally) depend on it.

**Comment by maxirmx** [comment]
@ni4 @ronaldtse @antonsviridenko @ribose-jeffreylau 
I am seeking 
- review of the methods used to incorporate sexp library into rnp
- advice how to update fuzz project build settings
- advice how to publish sexp to nix channel
- any other suggestions notes or compains

This PR is farily big already and includes mainly integration of the library but does not include minor changes and tests of the new format. I beieve furthe work shall go to the next PR.

Please do not merge it even if you like it. It waits for for https://github.com/rnpgp/sexp/pull/15

**Comment by ribose-jeffreylau** [comment]
@maxirmx Regarding publishing to nixpkgs (I assume that's what you meant by "nix channel"), the actual process is fairly simple.  The gist is that you create a PR against [github:NixOS/nixpkgs](https://github.com/NixOS/nixpkgs) with the new derivation file, and when it's merged (after going through reviews), the job is done.  [The long version can be found here.](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md)

This is not urgent as we also have Nix Flakes enabled for sexp (from your PR, thanks!), which means if another software (that also uses Nix Flakes) wants to include sexp, it can simply refer to `github:rnpgp/sexp` (and optionally the commit SHA) as one of its flake inputs, and it'll have access to the sexp build outputs (cli, libs and includes).

**Comment by maxirmx** [comment]
https://github.com/google/oss-fuzz/pull/9092

**Comment by ni4** [comment]
@maxirmx JFYI, I'm checking this it just takes some time. Not ignoring :)

**Comment by maxirmx** [comment]
@ni4 , thank you
I think this PR is агддн "ready for review"

LGTM reached end of life
https://github.blog/2022-08-15-the-next-step-for-lgtm-com-github-code-scanning/

**Comment by ni4** [comment]
> LGTM reached end of life https://github.blog/2022-08-15-the-next-step-for-lgtm-com-github-code-scanning/
@maxirmx Yeah, it seems we should migrate to GitHub code scanning. Anyway, it still seem to work and doesn't build due to CMake error:
```
[2022-12-01 16:08:57] [build-stdout] -- Configuring incomplete, errors occurred!
[2022-12-01 16:08:57] [build-stderr] CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
[2022-12-01 16:08:57] [build-stderr] Please set them or make sure they are set and tested correctly in the CMake files:
[2022-12-01 16:08:57] [build-stderr] /opt/src/src/lib/SEXP_INCLUDE_DIR
```

**Comment by maxirmx** [comment]
@ni4 **16th of December: LGTM.com will be shut down** 
Do you really think there is any value to fix it ?

**Comment by ronaldtse** [comment]
We seem to have a 10 day window until Dec 16 when LGTM shuts down, so let's quickly merge this when approved. Thanks!

File: src/librekey/g10_sexp.hpp
Comment (ni4) [inline src/librekey/g10_sexp.hpp:56]: It would be better to use some g10-specific C++-style naming, like G10SExp, but that could be done as a separate task.

**Review by ni4 — COMMENTED** [review]
Please see the comments. The main concern - should we support somehow people who don't need G10 support at all? This can be done as the separate task to not overbloat this PR.
cc @ronaldtse

File: src/lib/CMakeLists.txt
Comment (maxirmx) [reply src/lib/CMakeLists.txt:411]: If a client application uses shared rnp library, sexp is statically linked to librnp.so
If a client application uses static rnp library, it needs libsexp.a.

File: ci/lib/install_functions.inc.sh
Comment (maxirmx) [reply ci/lib/install_functions.inc.sh:983]: OK

File: cmake/Modules/DownloadSexp.cmake
Comment (maxirmx) [reply cmake/Modules/DownloadSexp.cmake:62]: OK

File: docs/installation.adoc
Comment (maxirmx) [reply docs/installation.adoc:240]: Sure, I missed gihub.com :) Thank you

File: src/librekey/g10_sexp.hpp
Comment (maxirmx) [reply src/librekey/g10_sexp.hpp:56]: Well, it was the integration point. I left class names and their interfaces unchanged for ```s_exp_t```, ```s_exp_elemement_t``` and put there the new implementation.  I would rather keep it as is within the scope of this specific PR

**Comment by maxirmx** [comment]
@ni4 Thank you for the comments
Kindly look at my fixes and replies

**Comment by maxirmx** [comment]
> please see the comments. The main concern - should we support somehow people who don't need G10 support at all? This can be done as the separate task to not overbloat this PR.
> cc @ronaldtse

IMHO the objective was this PR was to replace SEXP parser with a more generic one from a separate (static) library
I fixed several other minor issues 'on the go' but I think it is too risky to add more features.

The next PR is supposed to add tests for gnupg 2.3 imports. I can probably implement some other features or refacroing as well.

**Review by ni4 — APPROVED** [review]
LGTM, thanks! Please squash 'review fixes' commit if it is possible and feel free to merge afterwards.

**Pull request merged**
